### DISCUSSION TITLE
How to delete a node on a button click?

### DISCUSSION THREAD

**Discussion by clackd1amond** [xyflow/xyflow#1143] [Q&A] [answered]
Hi. I'd like to add a feature of deleting a node with a button that is a part of this node's label component. How do I do that?

**Comment by gilneto8** [comment]
Hey @clackd1amond,

I have a similar feature on my application, where we use a cross over the node to remove that node from the flow. What we do is we use the `removeElements` function to delete an array of nodes from the available ones. In this case, since we're building for the same functionality, I'm guessing your code would end up looking somewhat like this:

```javascript
import { removeElements } from 'react-flow-renderer';

// ...
const [elements, setElements] = useState([]);
const removeNode = (node) => {
   setElements((els) => removeElements([node], els));
}

// ...

// and then on your component

// ...

<button onClick={() => removeNode(node)}>Delete</button>
```

Hope to have helped.

Gil

**Comment by brenolapreza** [comment]
Hi,

Did you manage to solve this problem? If yes, could you share? Tks!

**Reply by mmexploit** [reply]
> Hi,
> 
> Did you manage to solve this problem? If yes, could you share? Tks!

Hello,
I thought of just using a filter function that gets called with the specific id. Haven't used the removeElements function, anyone who has?

**Comment by syntrydy** [comment]
> Hi. I'd like to add a feature of deleting a node with a button that is a part of this node's label component. How do I do that?

you can fire the below method on your onClick event. pass in the node Id

```js
const deleteNodeById = (id) => {
    flowInstance.setNodes((nds) => nds.filter((node) => node.id !== id))
 } 
```

**Comment by navaniharsh31** [comment, marked as answer]
you can use the `deleteElements` function given by react flow if youre on or above v11.2

https://reactflow.dev/docs/api/react-flow-instance/#deleteelements

**Reply by monil80** [reply]
Yes but how to implement it  I want to delete custom node with one btn click this button also in custom node

**Reply by chirag-thirstydevs** [reply]
I think you can pass node Id on click event and call this function to delete it.

**Reply by bcakmakoglu** [reply]
@monil80 
[Here's an example.](https://codesandbox.io/s/currying-fog-9jwtyy?file=/src/CustomNode.tsx)

1. Wrap your Flow component in `ReactFlowProvider` to get access to the `useReactFlow` hook
2. Call `useReactFlow` in your custom node, which will give you access to `deleteElements`
3. Call `deleteElements({ nodes: [...] })` passing whatever nodes you want to delete - it is enough to pass an obj like `{ id: '1' }` you don't need the whole node obj for this
4. Profit.

**Reply by monil80** [reply]
Hey thx for the help i saw this example but in my case custom node is separate file and my function is in main file  all i want  in dynamic way

**Reply by bcakmakoglu** [reply]
What do you mean "custom node is a separate file"?
Doesn't really matter *where* your custom node code is tbh, a separate file, the same file, both are fine.

You'll have to explain your case a bit, otherwise it's hard to guess what your issue is here ^^

**Comment by monil80** [comment]
> @monil80 [Here's an example.](https://codesandbox.io/s/currying-fog-9jwtyy?file=/src/CustomNode.tsx)

**Comment by chirag-thirstydevs** [comment]
* {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
        }
    



   Hello,Braks I hope you are having a great experience from our support agent .We would love to hear honest feedback on your experience on support for ticket #PL-8664913.Share your review here

**Comment by chirag-thirstydevs** [comment]
* {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
        }
    



   Hello,Braks I hope you are having a great experience from our support agent .We would love to hear honest feedback on your experience on support for ticket #PL-4241029.Share your review here

**Comment by chirag-thirstydevs** [comment]
* {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
        }
    



   Hello,monil I hope you are having a great experience from our support agent .We would love to hear honest feedback on your experience on support for ticket #PL-8435956.Share your review here

**Comment by i-m-afk** [comment]
If you are keeping a global store for nodes using "zustand", you can do something like this
```js
removeNode: (nodeId) => {
    set((state) => ({
      nodes: state.nodes.filter((node) => node.id !== nodeId),
      edges: state.edges.filter(
        (edge) => edge.source !== nodeId && edge.target !== nodeId,
      ),
    }));
  },
  ```
  
 then this in your node that you want to delete 
 ```js
  const removeNode = useStore((state) => state.removeNode);

  const handleDelete = useCallback(() => {
    removeNode(id);
  }, [id, removeNode]);

```
Within your node's JSX
  ```jsx
<button onClick={handleDelete}>Delete Node</button>
```
